README.md
[![Build Status](https://travis-ci.org/nunux-keeper/node-keeper.svg?branch=master)](https://travis-ci.org/nunux-keeper/node-keeper)

node-keeper
===========

Official <a href="http://keeper.nunux.org" target="_new">Nunux Keeper V2</a>
Node.js client.

Please note that this client is not compatible with Nunux Keeper v1.

Install
-------

Run the following command in the root directory of your project

```
npm install --save node-keeper
```

Basic Usage
-----------

>> See a complete working example in the `examples` directory.

```javascript
const express = require('express')
const KeeperSDK = require('node-keeper')

// YOUR persitence lib
const CredentialsDAO = require(/*???*/)

const sdk = new KeeperSDK({debug: true})

const options = {
  clientId: /** YOUR CLIENT ID */,
  clientSecret: /** YOUR CLIENT SECRET **/
}
const callback_url = /**YOUR CALLBACK URL **/

const app = express()

// Login entrypoint
app.get('/login', function(req, res) {
  sdk.createClient(options)
  .then((client) => {
    res.redirect(client.getAuthorizeURL(callback_url))
  })
})

// Callback URL
app.get('/callback', function(req, res) {
  if (req.query.error) {
    return res.send('ERROR: ' + req.query.error + ': ' + req.query.error_description)
  }

  sdk.createClient(options)
  .then((client) => {
    client.getTokens(callback_url, req.query.code)
    .then((creds) => {
      // Save credentials for further usage...
      CredentialsDAO.save(creds)
      res.send({ message: 'Authenticated :)' })
    })
    .catch((err) => {
      res.status(502).send({ error: 'something blew up', detail: err })
    })
  })
})

// Get protected resource
app.get('/document/:id', function(req, res) {
  // Init. client.
  sdk.createClient(Object.assign(options, {
    onRefreshTokens: (creds) => {
      // Save credentials for further usage...
      CredentialsDAO.save(creds)
    }
  }), CredentialsDAO.get())
  .then(function (client) {
    const result = client.api.getDocument(req.params.id)
    result.then((data) => {
      res.send(data)
    }).catch((err) => {
      res.status(502).send({ error: 'something blew up', detail: err })
    })
  })
})


app.listen(3000)
```

Copyright and License
---------------------

Copyright 2016 Nunux, Org. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
Changelog
=========

0.2.0
-----

### Features

* Update API definition (breaks compatibility with 0.1.x)

0.1.0
-----

### Features

* Refactoring (breaks compatibility with 0.0.x)
* Use API endpoint to discover authentication URL
* Create working example
* Add tests
* Add profile API

0.0.1
-----

### Features

* Get access token from credentials
* Create/Update/Delete/Retrieve a document
* Refreshing permanently access token with the refresh token
* Extract display name from the token



examples/app-auth/index.js
// -----------------------------------------------------------------------------
// Requirements
// -----------------------------------------------------------------------------
const express = require('express')
const exphbs = require('express-handlebars')
const session = require('express-session')
const KeeperSDK = require('node-keeper')

// ------------------------------------------------------------------------------
// Application Parameters - Fill in with your app's values
// ------------------------------------------------------------------------------

const CLIENT_ID = 'YOUR CLIENT ID'
const CLIENT_SECRET = 'YOUR CLIENT SECRET'

// Set up Express
const app = express()

const sdk = new KeeperSDK({debug: true})

const options = {
  clientId: CLIENT_ID,
  clientSecret: CLIENT_SECRET,
  offline: false
}

const redirect_uri = 'http://home:3000/callback'

// Set up the templating engine (Handlebars)
app.engine('hbs', exphbs({
  defaultLayout: 'main',
  extname: '.hbs'
}))
app.set('view engine', 'hbs')

// Set up sessions, so we can log users in and out
app.use(session({
  secret: 'session secret',
  resave: false,
  saveUninitialized: false
}))

// User authentication middleware
app.use(function (req, res, next) {
  if (req.session.credentials) {
    res.locals.credentials = req.session.credentials
  }
  next()
})

app.get('/', function (req, res) {
  // ender the home page
  res.render('home')
})

app.get('/login', function (req, res) {
  sdk.createClient(options)
  .then(function (client) {
    res.redirect(client.getAuthorizeURL(redirect_uri))
  })
})

app.get('/logout', function (req, res) {
  // To log the user out, we can simply destroy their session
  req.session.destroy(function () {
    res.redirect('/')
  })
})

// URL called back from the authetication service
app.get('/callback', function (req, res) {
  if (req.query.error) {
    return res.status(500).render('error', {
      error: req.query.error,
      cause: req.query.error_description
    })
  }

  sdk.createClient(options)
  .then(function (client) {
    client.getTokens(redirect_uri, req.query.code)
    .then((creds) => {
      // Save credentials for further usage...
      req.session.credentials = creds
      return res.redirect('/profile')
    })
    .catch((err) => {
      res.status(502).render('error', {
        error: 'Somethin blew up!',
        cause: err
      })
    })
  })
})

// Get protected resource
app.get('/profile', function (req, res) {
  if (!req.session.credentials) {
    // Return to home page if user is not logged
    return res.redirect('/')
  }

  // Init. client.
  sdk.createClient(Object.assign(options, {
    onRefreshTokens: (creds) => req.session.credentials = creds
  }), req.session.credentials)
  .then(function (client) {
    const result = client.api.profile.get()
    result.then((data) => {
      res.render('profile', {profile: data})
    }).catch((err) => {
      res.status(502).render('error', {
        error: 'Somethin blew up!',
        cause: err
      })
    })
  })
})

app.listen(3000)
console.log('Server started!')
console.log('Visit http://localhost:3000/ to start.')
